rs/nervous_system/candid_utils/src/validation.rs

```diff
@@ -70,33 +70,54 @@ impl CandidServiceArgValidationError {
     }
 }
 
+impl std::fmt::Display for CandidServiceArgValidationError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let (kind, message) = self.deconstruct();
+        write!(f, "CandidServiceArgValidationError::{kind}: {message}")
+    }
+}
+
 /// Checks whether `upgrade_args` is a valid argument sequence for `candid_service`.
 ///
-/// Returns the byte encoding of `upgrade_args` in the successful case.
+/// If `upgrade_args` is None, checks that `candid_service` does not require any arguments.
+///
+/// Example `upgrade_args` with two arguments: "((42 : nat32), opt record { foo = opt bar; })".
+///
+/// Returns the byte encoding of `upgrade_args` (if any; otherwise None) in the successful case.
 pub fn validate_upgrade_args(
     candid_service: String,
-    upgrade_args: String,
-) -> Result<Vec<u8>, CandidServiceArgValidationError> {
+    upgrade_args: Option<String>,
```

Comment:
Is "()" equivalent to None? If so, maybe we should just require the caller to pass "()" instead of None.

In general, I think we should avoid None when it is equivalent to some standard value (which is often just T::default()). None should be reserved for indicating ABSENCE of data, not presence of "boring" data.

Reply:
Our canister APIs are designed with `Option<Vec<u8>>` as the arg type in most cases.

For that reason, I make this function take the same type, so the clients don't need to think about how to process the "no arguments for this service" situation (as it's handles by the library).